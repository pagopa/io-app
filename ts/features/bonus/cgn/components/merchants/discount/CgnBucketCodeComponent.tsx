import { TouchableWithoutFeedback } from "@gorhom/bottom-sheet";
import { Millisecond } from "@pagopa/ts-commons/lib/units";
import * as React from "react";
import { useEffect, useRef } from "react";
import { View, Alert, StyleSheet } from "react-native";
import { Icon, VSpacer } from "@pagopa/io-app-design-system";
import { Discount } from "../../../../../../../definitions/cgn/merchants/Discount";
import { InfoBox } from "../../../../../../components/box/InfoBox";
import { BaseTypography } from "../../../../../../components/core/typography/BaseTypography";
import { H3 } from "../../../../../../components/core/typography/H3";
import { H4 } from "../../../../../../components/core/typography/H4";
import { IOStyles } from "../../../../../../components/core/variables/IOStyles";
import ActivityIndicator from "../../../../../../components/ui/ActivityIndicator";
import I18n from "../../../../../../i18n";
import { useIODispatch, useIOSelector } from "../../../../../../store/hooks";
import { clipboardSetStringWithFeedback } from "../../../../../../utils/clipboard";
import {
  isError,
  isLoading,
  isReady
} from "../../../../../../common/model/RemoteValue";
import {
  cgnCodeFromBucket,
  cgnCodeFromBucketReset
} from "../../../store/actions/bucket";
import { cgnBucketSelector } from "../../../store/reducers/bucket";
import { isDiscountBucketCodeResponseSuccess } from "../../../types/DiscountBucketCodeResponse";

type Props = {
  discountId: Discount["id"];
  onCodePress: (eventName: string) => void;
};

const styles = StyleSheet.create({
  row: {
    flexDirection: "row"
  },
  codeContainer: { alignItems: "center", justifyContent: "space-between" },
  codeText: {
    fontSize: 20
  },
  flexEnd: { alignSelf: "flex-end" }
});

const COPY_ICON_SIZE = 24;
const FEEDBACK_TIMEOUT = 3000 as Millisecond;
const EMPTY_CODE_CONTENT = "••••••••••";
type ContentProps = {
  onRequestBucket: () => void;
};

const BucketCodeHandler = ({
  onPress,
  content,
  icon
}: {
  onPress: () => void;
  content: string;
  icon: React.ReactElement;
}) => (
  <TouchableWithoutFeedback
    onPress={onPress}
    accessible={true}
    accessibilityRole={"button"}
    accessibilityHint={I18n.t("bonus.cgn.accessibility.code")}
  >
    <View style={[styles.row, styles.codeContainer]}>
      <BaseTypography
        weight={"Bold"}
        color={"bluegreyDark"}
        font={"RobotoMono"}
        style={styles.codeText}
      >
        {content}
      </BaseTypography>

      {icon}
    </View>
  </TouchableWithoutFeedback>
);
const CgnBucketCodeContent = (props: ContentProps) => {
  const [isTap, setIsTap] = React.useState(false);
  const timerRetry = useRef<number | undefined>(undefined);

  const bucketResponse = useIOSelector(cgnBucketSelector);

  const handleCopyPress = () => {
    if (
      isReady(bucketResponse) &&
      isDiscountBucketCodeResponseSuccess(bucketResponse.value)
    ) {
      setIsTap(true);
      clipboardSetStringWithFeedback(bucketResponse.value.value.code);
      // eslint-disable-next-line functional/immutable-data
      timerRetry.current = setTimeout(() => setIsTap(false), FEEDBACK_TIMEOUT);
    }
  };

  useEffect(
    () => () => {
      if (timerRetry.current) {
        clearTimeout(timerRetry.current);
      }
    },
    []
  );

  const showAlertError = () =>
    Alert.alert(
      I18n.t(
        "bonus.bonusVacanze.eligibility.activateBonus.discrepancies.attention"
      ),
      I18n.t("bonus.cgn.otp.error"),
      [
        {
          text: I18n.t("bonus.cgn.merchantDetail.bucket.alert.cta"),
          style: "cancel"
        }
      ]
    );

  useEffect(() => {
    if (
      isError(bucketResponse) ||
      (isReady(bucketResponse) && bucketResponse.value.kind === "unhandled")
    ) {
      showAlertError();
    }
  }, [bucketResponse]);

  if (isLoading(bucketResponse)) {
    return <ActivityIndicator />;
  }

  // we got an error no code is available
  if (isReady(bucketResponse) && bucketResponse.value.kind === "notFound") {
    return (
      <>
        <InfoBox
          iconColor="aqua"
          iconName="errorFilled"
          iconSize={24}
          alignedCentral
        >
          <H4 weight={"Regular"} style={IOStyles.flex}>
            {I18n.t("bonus.cgn.merchantDetail.bucket.error.noCode")}
          </H4>
        </InfoBox>
      </>
    );
  }

  if (isReady(bucketResponse) && bucketResponse.value.kind === "success") {
    return (
      <BucketCodeHandler
        onPress={handleCopyPress}
        content={bucketResponse.value.value.code}
        icon={
          <View style={styles.flexEnd}>
            <Icon
              name={isTap ? "ok" : "copy"}
              size={COPY_ICON_SIZE}
              color="blue"
            />
          </View>
        }
      />
    );
  }

  return (
    <BucketCodeHandler
      onPress={props.onRequestBucket}
      content={EMPTY_CODE_CONTENT}
      icon={<Icon size={COPY_ICON_SIZE} name="eyeShow" color="blue" />}
    />
  );
};
const CgnBucketCodeComponent = ({ discountId, onCodePress }: Props) => {
  const dispatch = useIODispatch();

  const requestBucketCode = () => {
    onCodePress("CGN_BUCKET_CODE_START_REQUEST");
    dispatch(cgnCodeFromBucket.request(discountId));
  };
  useEffect(
    () => () => {
      dispatch(cgnCodeFromBucketReset());
    },
    [dispatch]
  );

  return (
    <View testID={"bucket-code-component"}>
      <H3 accessible={true} accessibilityRole={"header"}>
        {I18n.t("bonus.cgn.merchantDetail.title.discountCode")}
      </H3>
      <VSpacer size={8} />
      <CgnBucketCodeContent onRequestBucket={requestBucketCode} />
    </View>
  );
};

export default CgnBucketCodeComponent;
