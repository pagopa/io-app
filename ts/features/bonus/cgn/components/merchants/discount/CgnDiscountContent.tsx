import React from "react";
import {
  ContentWrapper,
  Divider,
  ListItemInfo
} from "@pagopa/io-app-design-system";
import { Discount } from "../../../../../../../definitions/cgn/merchants/Discount";
import I18n from "../../../../../../i18n";
import { localeDateFormat } from "../../../../../../utils/locale";

type CgnDiscountContentProps = {
  discountDetails: Discount;
};

export const CgnDiscountContent = ({
  discountDetails
}: CgnDiscountContentProps) => (
  <ContentWrapper>
    {discountDetails.description && (
      <>
        <ListItemInfo
          numberOfLines={0}
          label={I18n.t("bonus.cgn.merchantDetail.discount.description")}
          value={discountDetails.description}
        />
        <Divider />
      </>
    )}
    {discountDetails.condition && (
      <>
        <ListItemInfo
          numberOfLines={0}
          label={I18n.t("bonus.cgn.merchantDetail.discount.conditions")}
          value={discountDetails.condition}
        />
        <Divider />
      </>
    )}
    {discountDetails.startDate && discountDetails.endDate && (
      <ListItemInfo
        numberOfLines={0}
        label={I18n.t("bonus.cgn.merchantDetail.discount.validity")}
        value={`${localeDateFormat(
          discountDetails.startDate,
          I18n.t("global.dateFormats.shortFormat")
        )} - ${localeDateFormat(
          discountDetails.endDate,
          I18n.t("global.dateFormats.shortFormat")
        )}`}
      />
    )}
  </ContentWrapper>
);
