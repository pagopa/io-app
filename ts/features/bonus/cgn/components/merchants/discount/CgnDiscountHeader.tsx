import * as E from "fp-ts/lib/Either";
import React from "react";
import { StyleSheet, View } from "react-native";
import {
  Badge,
  H3,
  IOColors,
  IOStyles,
  VSpacer
} from "@pagopa/io-app-design-system";
import { useHeaderHeight } from "@react-navigation/elements";
import { pipe } from "fp-ts/lib/function";
import { WithinRangeInteger } from "@pagopa/ts-commons/lib/numbers";
import { Discount } from "../../../../../../../definitions/cgn/merchants/Discount";
import { CategoryTag } from "../CgnModuleDiscount";
import I18n from "../../../../../../i18n";

type CgnDiscountHeaderProps = {
  onLayout: (event: any) => void;
  discountDetails: Discount;
};

export const CgnDiscountHeader = ({
  onLayout,
  discountDetails
}: CgnDiscountHeaderProps) => {
  const discountColor = discountDetails.isNew
    ? styles.backgroundNewItem
    : styles.backgroundDefault;

  const normalizedDiscountValue = pipe(
    WithinRangeInteger(0, 100).decode(discountDetails.discount),
    E.map(v => v.toString()),
    E.getOrElse(() => "-")
  );

  const headerHeight = useHeaderHeight();

  return (
    <View
      onLayout={onLayout}
      style={[
        IOStyles.horizontalContentPadding,
        {
          paddingTop: headerHeight,
          backgroundColor: discountColor.backgroundColor,
          paddingBottom: 24
        }
      ]}
    >
      <View>
        {(discountDetails.isNew || discountDetails.discount) && (
          <>
            <View style={[IOStyles.row, { gap: 8 }]}>
              {discountDetails.isNew && (
                <Badge
                  variant="purple"
                  text={I18n.t("bonus.cgn.merchantsList.news")}
                />
              )}
              {discountDetails.discount ? (
                <Badge
                  variant="purple"
                  outline
                  text={`-${normalizedDiscountValue}%`}
                />
              ) : null}
            </View>
            <VSpacer size={12} />
          </>
        )}
        <H3>{discountDetails.name}</H3>
        <VSpacer size={12} />
        <View style={[{ flexWrap: "wrap" }, IOStyles.row]}>
          {discountDetails.productCategories.map(categoryKey => (
            <CategoryTag key={categoryKey} category={categoryKey} />
          ))}
        </View>
      </View>
    </View>
  );
};

const styles = StyleSheet.create({
  backgroundDefault: {
    backgroundColor: IOColors["grey-50"],
    borderColor: IOColors["grey-100"]
  },
  backgroundNewItem: {
    backgroundColor: IOColors["hanPurple-50"],
    borderColor: IOColors["hanPurple-250"]
  }
});
