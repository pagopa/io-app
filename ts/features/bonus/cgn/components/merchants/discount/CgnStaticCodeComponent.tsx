import { TouchableWithoutFeedback } from "@gorhom/bottom-sheet";
import { Millisecond } from "@pagopa/ts-commons/lib/units";
import * as React from "react";
import { useCallback } from "react";
import { View, StyleSheet } from "react-native";
import { Icon } from "@pagopa/io-app-design-system";
import { Discount } from "../../../../../../../definitions/cgn/merchants/Discount";
import { BaseTypography } from "../../../../../../components/core/typography/BaseTypography";
import { H3 } from "../../../../../../components/core/typography/H3";
import I18n from "../../../../../../i18n";
import { clipboardSetStringWithFeedback } from "../../../../../../utils/clipboard";

type Props = {
  staticCode: Discount["staticCode"];
  onCodePress: (eventName: string) => void;
};

const styles = StyleSheet.create({
  row: {
    flexDirection: "row"
  },
  codeContainer: { alignItems: "center", justifyContent: "space-between" },
  codeText: {
    fontSize: 20
  },
  flexEnd: { alignSelf: "flex-end" }
});

const FEEDBACK_TIMEOUT = 3000 as Millisecond;
const COPY_ICON_SIZE = 24;

const CgnStaticCodeComponent: React.FunctionComponent<Props> = ({
  staticCode,
  onCodePress
}: Props) => {
  const [isTap, setIsTap] = React.useState(false);
  const [isCodeVisible, setIsCodeVisible] = React.useState(false);
  const timerRetry = React.useRef<number | undefined>(undefined);

  React.useEffect(
    () => () => {
      clearTimeout(timerRetry.current);
    },
    []
  );

  const handleCopyPress = useCallback(() => {
    if (staticCode) {
      setIsTap(true);
      clipboardSetStringWithFeedback(staticCode);
      // eslint-disable-next-line functional/immutable-data
      timerRetry.current = setTimeout(() => setIsTap(false), FEEDBACK_TIMEOUT);
    }
  }, [staticCode]);

  const requestStaticCode = useCallback(() => {
    onCodePress("CGN_STATIC_CODE_REQUEST");
    setIsCodeVisible(true);
  }, [onCodePress]);

  return (
    <>
      <H3 accessible={true} accessibilityRole={"header"}>
        {I18n.t("bonus.cgn.merchantDetail.title.discountCode")}
      </H3>
      <TouchableWithoutFeedback
        onPress={isCodeVisible ? handleCopyPress : requestStaticCode}
        accessible={true}
        accessibilityRole={"button"}
        accessibilityHint={I18n.t("bonus.cgn.accessibility.code")}
        testID={"static-code-component"}
      >
        <View style={[styles.row, styles.codeContainer]}>
          <BaseTypography
            weight={"Bold"}
            color={"bluegreyDark"}
            font={"RobotoMono"}
            style={styles.codeText}
          >
            {isCodeVisible && staticCode ? staticCode : "••••••••••"}
          </BaseTypography>

          {isCodeVisible ? (
            <View style={styles.flexEnd}>
              <Icon
                name={isTap ? "ok" : "copy"}
                size={COPY_ICON_SIZE}
                color="blue"
              />
            </View>
          ) : (
            <Icon name="eyeShow" size={COPY_ICON_SIZE} color="blue" />
          )}
        </View>
      </TouchableWithoutFeedback>
    </>
  );
};
export default CgnStaticCodeComponent;
