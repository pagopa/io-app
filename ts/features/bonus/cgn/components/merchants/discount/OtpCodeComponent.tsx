import { Millisecond, Second } from "@pagopa/ts-commons/lib/units";
import { pipe } from "fp-ts/lib/function";
import * as O from "fp-ts/lib/Option";
import React, { useEffect, useRef, useState } from "react";
import { Animated, Easing, StyleSheet, View, ViewStyle } from "react-native";
import { IOColors, Icon } from "@pagopa/io-app-design-system";
import { Otp } from "../../../../../../../definitions/cgn/Otp";
import { BaseTypography } from "../../../../../../components/core/typography/BaseTypography";
import { H5 } from "../../../../../../components/core/typography/H5";
import TouchableDefaultOpacity from "../../../../../../components/TouchableDefaultOpacity";
import I18n from "../../../../../../i18n";
import { clipboardSetStringWithFeedback } from "../../../../../../utils/clipboard";
import { isTestEnv } from "../../../../../../utils/environment";
import { addEvery } from "../../../../../../utils/strings";

type ProgressConfig = {
  startPercentage: number;
  endPercentage: number;
};

type Props = {
  otp: Otp;
  progressBaseBgColor?: ViewStyle["backgroundColor"];
  progressBgColor?: ViewStyle["backgroundColor"];
  onEnd: () => void;
  progressConfig?: ProgressConfig;
};

const styles = StyleSheet.create({
  container: {
    backgroundColor: IOColors.white
  },
  otpContainer: {
    marginVertical: 4,
    justifyContent: "space-between",
    backgroundColor: IOColors.white,
    flexDirection: "row"
  },
  progressBase: {
    alignSelf: "center",
    marginTop: 10,
    marginBottom: 10,
    backgroundColor: IOColors.bluegreyLight,
    width: "100%",
    height: 2,
    margin: 4,
    borderRadius: 4
  },
  progress: {
    alignSelf: "flex-start",
    backgroundColor: IOColors.blue,
    width: "50%",
    height: 2,
    borderRadius: 4
  },
  remainingTimeContainer: {
    alignSelf: "center",
    width: "100%"
  },
  optCode: { fontSize: 16, textAlign: "left" }
});

// Monospace custom component
const OtpCode = (code: string) => (
  <BaseTypography
    color={"bluegrey"}
    weight={"Regular"}
    fontStyle={styles.optCode}
    font={"RobotoMono"}
  >
    {code}
  </BaseTypography>
);

type ExpirationTime = { minutes: number; seconds: number };

/**
 * compute otp code expiration time
 * - if the otp elapsed time is before the current time, the remaining time is ttl
 * - if the otp elapsed time is after the current time, the remaining time is the minimum between now - elapsed time and ttl
 */
const getOtpTTL = (otp: Otp): Millisecond => {
  const now = new Date();
  const expiration = otp.expires_at.getTime() - now.getTime();
  if (expiration > 0) {
    // take the min between ttl and computed seconds
    return Math.min(Math.ceil(expiration), otp.ttl * 1000) as Millisecond;
  }
  // expires is in the past relative to the dice current time, use ttl as fallback
  return (otp.ttl * 1000) as Millisecond;
};

// return a string (or undefined) of the remaining time
const getRemainingTimeRepr = (
  expirationtime: ExpirationTime
): string | undefined =>
  pipe(
    expirationtime,
    O.fromNullable,
    O.map(rt => {
      const minutes =
        rt.minutes === 0
          ? ""
          : I18n.t("bonus.cgn.otp.code.minutes", {
              defaultValue: I18n.t("bonus.cgn.otp.code.minutes.other", {
                minutes: rt.minutes
              }),
              count: rt.minutes,
              minutes: rt.minutes
            }) + " ";
      const seconds = I18n.t("bonus.cgn.otp.code.seconds", {
        defaultValue: I18n.t("bonus.cgn.otp.code.seconds.other", {
          seconds: rt.seconds
        }),
        count: rt.seconds,
        seconds: rt.seconds
      });
      return `${I18n.t("bonus.cgn.otp.code.validUntil")} ${minutes}${seconds}`;
    }),
    O.toUndefined
  );

export const OtpCodeComponent = (props: Props) => {
  const { startPercentage, endPercentage } = props.progressConfig ?? {
    startPercentage: 0,
    endPercentage: 100
  };
  const [progressWidth] = useState(new Animated.Value(startPercentage));
  const [elapsedSeconds, setElapsedSeconds] = useState(0);
  const [duration, setDuration] = useState<Second>(0 as Second);
  const [expirationTime, setExpirationTime] = useState<
    ExpirationTime | undefined
  >(undefined);
  const intervalRef = useRef<undefined | number>(undefined);
  const formattedCode = addEvery(props.otp.code, " ", 3);

  // reset interval timer
  const stopInterval = () => clearInterval(intervalRef.current);

  const { otp, onEnd } = props;

  /**
   * when otp changes
   * - reset elapsed seconds
   * - reset progress width
   * - start progress width animation
   */
  useEffect(() => {
    const currentDuration = getOtpTTL(otp);
    setDuration((currentDuration / 1000) as Second);
    setElapsedSeconds(0);
    progressWidth.setValue(startPercentage);

    Animated.timing(progressWidth, {
      useNativeDriver: false,
      toValue: endPercentage,
      duration: currentDuration as number,
      easing: Easing.linear
    }).start(({ finished }) => {
      // when animation end
      if (finished) {
        setElapsedSeconds(cv => cv + 1);
        stopInterval();
        onEnd();
      }
    });
    // eslint-disable-next-line functional/immutable-data
    intervalRef.current = setInterval(() => {
      setElapsedSeconds(cv => cv + 1);
    }, 1000);
    return stopInterval;
  }, [otp, onEnd, progressWidth, startPercentage, endPercentage]);

  // update the remaining time
  useEffect(() => {
    if (duration === 0) {
      return;
    }
    const expiration = duration - elapsedSeconds;
    const minutes = Math.max(0, Math.floor(expiration / 60));
    const seconds = Math.max(0, Math.floor(expiration % 60));
    setExpirationTime({ minutes, seconds });
  }, [duration, elapsedSeconds]);

  return (
    <View style={styles.container}>
      <TouchableDefaultOpacity
        style={styles.otpContainer}
        onPress={() => clipboardSetStringWithFeedback(props.otp.code)}
        accessible={true}
        accessibilityRole={"button"}
        accessibilityHint={I18n.t("bonus.cgn.accessibility.code")}
      >
        {OtpCode(formattedCode)}
        <View style={{ justifyContent: "center" }}>
          <Icon name="copy" color="blue" />
        </View>
      </TouchableDefaultOpacity>
      <View
        style={[
          styles.progressBase,
          {
            backgroundColor:
              props.progressBaseBgColor ?? styles.progressBase.backgroundColor
          }
        ]}
      >
        <Animated.View
          style={[
            styles.progress,
            {
              backgroundColor:
                props.progressBgColor ?? styles.progress.backgroundColor
            },
            {
              width: progressWidth.interpolate({
                inputRange: [0, 100],
                outputRange: ["0%", "100%"]
              })
            }
          ]}
        />
      </View>
      {expirationTime && (
        <View style={styles.remainingTimeContainer}>
          <H5 weight={"Regular"} color={"bluegrey"}>
            {getRemainingTimeRepr(expirationTime)}
          </H5>
        </View>
      )}
    </View>
  );
};

// keep encapsulation strong
export const testableOtpCodeRefreshComponent = isTestEnv
  ? { getOtpTTL }
  : undefined;
