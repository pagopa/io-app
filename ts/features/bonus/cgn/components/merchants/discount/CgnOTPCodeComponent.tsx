import * as React from "react";
import { TouchableWithoutFeedback } from "@gorhom/bottom-sheet";
import { View, StyleSheet } from "react-native";
import { Icon, LabelLink } from "@pagopa/io-app-design-system";
import I18n from "../../../../../../i18n";
import { BaseTypography } from "../../../../../../components/core/typography/BaseTypography";
import { useIODispatch, useIOSelector } from "../../../../../../store/hooks";
import { cgnOtpDataSelector } from "../../../store/reducers/otp";
import {
  isError,
  isLoading,
  isReady
} from "../../../../../../common/model/RemoteValue";
import { cgnGenerateOtp, resetOtpState } from "../../../store/actions/otp";
import ActivityIndicator from "../../../../../../components/ui/ActivityIndicator";
import { IOStyles } from "../../../../../../components/core/variables/IOStyles";
import { H3 } from "../../../../../../components/core/typography/H3";
import { H4 } from "../../../../../../components/core/typography/H4";
import { OtpCodeComponent } from "./OtpCodeComponent";

const styles = StyleSheet.create({
  row: {
    flexDirection: "row"
  },
  codeContainer: { alignItems: "center", justifyContent: "space-between" },
  codeText: {
    fontSize: 20
  }
});

const COPY_ICON_SIZE = 24;

type Props = {
  onCodePress: (eventName: string) => void;
};

const CgnOTPCodeContent = ({ onCodePress }: Props) => {
  const [isCodeVisible, setIsCodeVisible] = React.useState(false);
  const dispatch = useIODispatch();
  const otp = useIOSelector(cgnOtpDataSelector);

  const requestOtp = () => {
    onCodePress("CGN_OTP_START_REQUEST");
    dispatch(cgnGenerateOtp.request());
    setIsCodeVisible(true);
  };

  const resetOtp = () => {
    dispatch(resetOtpState());
    setIsCodeVisible(false);
  };

  if (isLoading(otp)) {
    return <ActivityIndicator />;
  }

  if (isCodeVisible) {
    if (isReady(otp)) {
      return (
        <OtpCodeComponent
          otp={otp.value}
          onEnd={resetOtp}
          progressConfig={{
            startPercentage: 100,
            endPercentage: 0
          }}
        />
      );
    }
    if (isError(otp)) {
      return (
        <TouchableWithoutFeedback
          onPress={requestOtp}
          accessible={true}
          accessibilityRole={"button"}
          accessibilityHint={I18n.t("bonus.cgn.accessibility.code")}
        >
          <View>
            <H4 weight={"Regular"} style={IOStyles.flex}>
              {I18n.t("bonus.cgn.otp.error")}
            </H4>

            <LabelLink>{I18n.t("global.buttons.retry")}</LabelLink>
          </View>
        </TouchableWithoutFeedback>
      );
    }
  }

  return (
    <TouchableWithoutFeedback
      onPress={requestOtp}
      accessible={true}
      accessibilityRole={"button"}
      accessibilityHint={I18n.t("bonus.cgn.accessibility.code")}
    >
      <View style={[styles.row, styles.codeContainer]}>
        <BaseTypography
          weight={"Bold"}
          color={"bluegreyDark"}
          font={"RobotoMono"}
          style={styles.codeText}
        >
          {"••••••••••"}
        </BaseTypography>

        <Icon size={COPY_ICON_SIZE} color="blue" name={"eyeShow"} />
      </View>
    </TouchableWithoutFeedback>
  );
};

const CgnOTPCodeComponent = (props: Props) => (
  <View testID={"otp-code-component"}>
    <H3 accessible={true} accessibilityRole={"header"}>
      {I18n.t("bonus.cgn.merchantDetail.title.discountCode")}
    </H3>
    <CgnOTPCodeContent {...props} />
  </View>
);

export default CgnOTPCodeComponent;
