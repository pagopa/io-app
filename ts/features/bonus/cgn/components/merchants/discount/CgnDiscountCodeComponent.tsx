import * as React from "react";
import { VSpacer } from "@pagopa/io-app-design-system";
import { Discount } from "../../../../../../../definitions/cgn/merchants/Discount";
import { DiscountCodeType } from "../../../../../../../definitions/cgn/merchants/DiscountCodeType";
import CgnStaticCodeComponent from "./CgnStaticCodeComponent";
import CgnOTPCodeComponent from "./CgnOTPCodeComponent";
import CgnBucketCodeComponent from "./CgnBucketCodeComponent";

type Props = {
  discount: Discount;
  onCodePress: (eventName: string) => void;
  merchantType?: DiscountCodeType;
};

const CgnDiscountCodeComponent = ({
  discount,
  merchantType,
  onCodePress
}: Props) => {
  const shouldNotRender =
    merchantType === "landingpage" || merchantType === undefined;
  const renderProperCodeVisualization = (discount: Discount) => {
    switch (merchantType) {
      case "api":
        return <CgnOTPCodeComponent onCodePress={onCodePress} />;
      case "static":
        return (
          <CgnStaticCodeComponent
            staticCode={discount.staticCode}
            onCodePress={onCodePress}
          />
        );
      case "bucket":
        return (
          <CgnBucketCodeComponent
            discountId={discount.id}
            onCodePress={onCodePress}
          />
        );
      case "landingpage":
      default:
        return null;
    }
  };

  return shouldNotRender ? null : (
    <>
      <VSpacer size={8} />
      {renderProperCodeVisualization(discount)}
      <VSpacer size={16} />
    </>
  );
};

export default CgnDiscountCodeComponent;
