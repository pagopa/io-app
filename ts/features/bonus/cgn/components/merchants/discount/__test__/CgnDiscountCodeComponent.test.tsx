import { render } from "@testing-library/react-native";
import * as React from "react";
import CgnDiscountCodeComponent from "../CgnDiscountCodeComponent";
import { DiscountCodeTypeEnum } from "../../../../../../../../definitions/cgn/merchants/DiscountCodeType";
import { mockDiscount } from "../../../../__mock__/discount";

const mockFunc = (_: string) => jest.fn();

describe("when rendering", () => {
  it("should match the snapshot", () => {
    expect(
      render(
        <CgnDiscountCodeComponent
          discount={mockDiscount}
          merchantType={DiscountCodeTypeEnum.static}
          onCodePress={mockFunc}
        />
      ).toJSON()
    ).toMatchSnapshot();
  });
});

describe("when should not render", () => {
  it("should match the snapshot", () => {
    expect(
      render(
        <CgnDiscountCodeComponent
          discount={mockDiscount}
          merchantType={DiscountCodeTypeEnum.landingpage}
          onCodePress={mockFunc}
        />
      ).toJSON()
    ).toMatchSnapshot();
  });
});
