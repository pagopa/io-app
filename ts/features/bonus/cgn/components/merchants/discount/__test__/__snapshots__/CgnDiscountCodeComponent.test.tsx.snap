// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when rendering should match the snapshot 1`] = `
Array [
  <View
    style={
      Object {
        "height": 8,
      }
    }
  />,
  <Text
    accessibilityRole="header"
    accessible={true}
    color="bluegreyDark"
    font="TitilliumWeb"
    fontStyle={
      Object {
        "fontSize": 18,
        "lineHeight": 22,
      }
    }
    style={
      Array [
        Object {
          "fontSize": 18,
          "lineHeight": 22,
        },
        Object {
          "color": "#17324D",
          "fontFamily": "Titillium Web",
          "fontStyle": "normal",
          "fontWeight": "600",
        },
      ]
    }
    weight="SemiBold"
    weightColorFactory={[Function]}
  >
    Discount code
  </Text>,
  <View
    accessibilityHint="By pressing the item you will copy the code"
    accessibilityRole="button"
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Array [
        Object {
          "flexDirection": "row",
        },
        Object {
          "alignItems": "center",
          "justifyContent": "space-between",
        },
      ]
    }
    testID="static-code-component"
  >
    <Text
      color="bluegreyDark"
      font="RobotoMono"
      style={
        Array [
          Object {
            "fontSize": 20,
          },
          undefined,
          Object {
            "color": "#17324D",
            "fontFamily": "Roboto Mono",
            "fontStyle": "normal",
            "fontWeight": "700",
          },
        ]
      }
      weight="Bold"
    >
      ••••••••••
    </Text>
    <RNSVGSvgView
      accessibilityElementsHidden={true}
      accessibilityLabel=""
      accessible={false}
      align="xMidYMid"
      bbHeight={24}
      bbWidth={24}
      color={4278219750}
      focusable={false}
      height={24}
      importantForAccessibility="no-hide-descendants"
      meetOrSlice={0}
      minX={0}
      minY={0}
      style={
        Array [
          Object {
            "backgroundColor": "transparent",
            "borderWidth": 0,
          },
          Object {
            "color": "#0073E6",
          },
          Object {
            "flex": 0,
            "height": 24,
            "width": 24,
          },
        ]
      }
      tintColor={4278219750}
      vbHeight={24}
      vbWidth={24}
      width={24}
    >
      <RNSVGGroup>
        <RNSVGPath
          clipRule={0}
          d="M2.377 11.667A10.997 10.997 0 0 1 12.001 6c4.138 0 7.744 2.285 9.623 5.666a.686.686 0 0 1 0 .667A10.997 10.997 0 0 1 12 18a10.997 10.997 0 0 1-9.623-5.666.686.686 0 0 1 0-.667zM12.001 4A12.997 12.997 0 0 0 .629 10.696a2.686 2.686 0 0 0 0 2.609A12.997 12.997 0 0 0 12 20c4.894 0 9.155-2.705 11.372-6.696.45-.812.45-1.797 0-2.609A12.997 12.997 0 0 0 12.001 4zM15 12a3 3 0 1 1-6 0 3 3 0 0 1 6 0zm2 0a5 5 0 1 1-10 0 5 5 0 0 1 10 0z"
          fill={
            Array [
              2,
            ]
          }
          fillRule={0}
          propList={
            Array [
              "fill",
              "fillRule",
            ]
          }
        />
      </RNSVGGroup>
    </RNSVGSvgView>
  </View>,
  <View
    style={
      Object {
        "height": 16,
      }
    }
  />,
]
`;

exports[`when should not render should match the snapshot 1`] = `null`;
